platform/android/java/lib/src/org/godotengine/godot/utils/PermissionsUtil.java

```diff
@@ -0,0 +1,70 @@
+package org.godotengine.godot.utils;
+
+import android.Manifest;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.support.v4.content.ContextCompat;
+import org.godotengine.godot.Godot;
+
+public final class PermissionsUtil {
+
+	static final int REQUEST_RECORD_AUDIO_PERMISSION = 1;
+	static final int REQUEST_CAMERA_PERMISSION = 2;
+	static final int REQUEST_VIBRATE_PERMISSION = 3;
+	static final int REQUEST_ALL_PERMISSION_REQ_CODE = 1001;
+
+	private PermissionsUtil() {
+	}
+
+	public static boolean requestPermission(String p_name, Godot p_activity) {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+			// Not necessary, asked on install already
+			return true;
+		}
+
+		if (p_name.equals("RECORD_AUDIO")) {
+			if (ContextCompat.checkSelfPermission(p_activity, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
+				p_activity.requestPermissions(new String[] { Manifest.permission.RECORD_AUDIO }, REQUEST_RECORD_AUDIO_PERMISSION);
+				return false;
+			}
+		}
+
+		if (p_name.equals("CAMERA")) {
+			if (ContextCompat.checkSelfPermission(p_activity, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
+				p_activity.requestPermissions(new String[] { Manifest.permission.CAMERA }, REQUEST_CAMERA_PERMISSION);
+				return false;
+			}
+		}
+
+		if (p_name.equals("VIBRATE")) {
+			if (ContextCompat.checkSelfPermission(p_activity, Manifest.permission.VIBRATE) != PackageManager.PERMISSION_GRANTED) {
+				p_activity.requestPermissions(new String[] { Manifest.permission.VIBRATE }, REQUEST_VIBRATE_PERMISSION);
+				return false;
+			}
+		}
+		return true;
+	}
+
+	public static boolean requestManifestPermissions(Godot p_activity) {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+			return true;
+		}
+
+		PackageInfo packageInfo;
+		try {
+			packageInfo = p_activity.getPackageManager().getPackageInfo(p_activity.getPackageName(),
+					PackageManager.GET_PERMISSIONS);
+		} catch (PackageManager.NameNotFoundException e) {
+			e.printStackTrace();
+			return false;
+		}
+
+		String[] manifestPermissions = packageInfo.requestedPermissions;
+		if (manifestPermissions == null || manifestPermissions.length == 0)
+			return true;
+
+		p_activity.requestPermissions(manifestPermissions, REQUEST_ALL_PERMISSION_REQ_CODE);
```

Comment:
Some feedbacks based on the documentation for [requestPermissions](https://developer.android.com/reference/androidx/core/app/ActivityCompat.html#requestPermissions(android.app.Activity,%20java.lang.String%5B%5D,%20int)):

- We should filter the list of permissions for `dangerous` only permissions. This part might be a bit difficult since there doesn't seem to be an official list of the `dangerous` only permissions, so we may have to add our own in `PermissionsUtil` based on sources like [this](https://github.com/aosp-mirror/platform_frameworks_base/blob/ea2bea3bedeeebce0b623d04546c2d0e6513bc91/core/res/AndroidManifest.xml#L664) or [this](https://stackoverflow.com/questions/36936914/list-of-android-permissions-normal-permissions-and-dangerous-permissions-in-api).
- With the filtered list, we need to again filter out the permissions that are already granted as we doing in the method above. If all permissions are granted, then we can return `true` right away.

In addition, we should move the logic for `onRequestPermissionsResult` in PermissionsUtil as well. Once the result for a permission request is available, in the native code, we should send a signal with a `true/false` argument to let the game logic know that the permission has been granted/denied.

As a helper method, we can add an additional `getGrantedPermissions` binding to let the game logic retrieves a list of strings representing the granted dangerous permissions.

Reply:
I check requested permission protection levels and only request for them. 

I commited something about `onRequestPermissionsResult` callback and send signal over `MainLoop`. Can you check [this](https://github.com/cagdasc/godot/commit/4b76081b8eece79d6e47d82d14b7883c83a4ca7f#diff-133edc4b3503a0551d94e51fe7b78ba8) and [this](https://github.com/cagdasc/godot/commit/4b76081b8eece79d6e47d82d14b7883c83a4ca7f#diff-7c53cc557c747dfb35393fd94faeef8d). If you want like this I can implement.

For `getGrantedPermissions`, I researched about that . I think we can get this list. If my last commit is okay, I will start develop this.

Reply:
The commits you linked looks good! So yes, let's use the same logic to emit a permission result signal when `onRequestPermissionsResult` is invoked.

For `getGrantedPermissions`, you should be able to use a logic similar to what you have in `PermissionsUtil.requestPermissions`, where you grab the list of all permissions, filter the dangerous ones, and only return the ones that are already granted.